### PR TITLE
MLP implementation 

### PR DESCRIPTION
**Pull request by kmd2zjw**

### CODE REVIEW

File: misc-bench/cpp-mlp/mlp.cpp [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+      std::cout << "Abort" << std::endl;
+      return;
+    }
+
+    status = pimAdd(srcObj2, dstObj, dstObj);
+    if (status != PIM_OK)
+    {
+      std::cout << "Abort" << std::endl;
+      return;
+    }
```
Comment (fasiddique) [inline misc-bench/cpp-mlp/mlp.cpp:173]: can you please update this with new gemv implementation?

**Review by fasiddique — CHANGES_REQUESTED** [review]

File: misc-bench/cpp-mlp/PIM/Makefile [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+EXEC := mlp.out
+SRC := mlp.cpp
+
+debug perf dramsim3_integ: $(EXEC)
+
+$(EXEC): $(SRC) $(DEPS)
+	$(CXX) $< $(CXXFLAGS) -o $@
+
+clean:
+	rm -rf $(EXEC) *.dSYM
```
Comment (deyuan) [inline misc-bench/cpp-mlp/PIM/Makefile:24]: Missing line break at the end of file

File: misc-bench/cpp-mlp/baselines/mlp.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+if __name__ == "__main__":
+    print("[INFO] Parsing command line arguments")
+    parser = argparse.ArgumentParser(description="runs an MLP inference, where the layers are defined by the user")
+    parser.add_argument("-cuda", action='store_true', help="Use GPU if available")
+    parser.add_argument("-l", default="500,1000,2", help="comma-seperate list of number the neurons in each layer (default=\"500,1000,2\")")
+    parser.add_argument("-n", type=int, default=1, help="number of inference points for the batch (default=1)")
+    args = parser.parse_args()
+    print("[INFO] Starting the inference process")
+    main(args)
```
Comment (deyuan) [inline misc-bench/cpp-mlp/baselines/mlp.py:105]: Missing line break at the end of file

File: misc-bench/cpp-mlp/PIM/mlp.cpp [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+#include <stdbool.h>
+#include <string.h>
+#include <unistd.h>
+#include <getopt.h>
+#include <assert.h>
+#include <stdint.h>
+
+#include "../../util.h"
+#include "libpimeval.h"
+using namespace std;
```
Comment (deyuan) [inline misc-bench/cpp-mlp/PIM/mlp.cpp:13]: One suggestion is to not using this, and explicitly write std:: in code

File: misc-bench/cpp-mlp/PIM/mlp.cpp [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+
+// Data type
+#define T int32_t
```
Comment (deyuan) [inline misc-bench/cpp-mlp/PIM/mlp.cpp:16]: Suggest a more comprehensive type name to ease searching

File: misc-bench/cpp-mlp/PIM/mlp.cpp [hunk 6]
```diff
@@ -0,0 +1,302 @@
+
```
Comment (deyuan) [inline misc-bench/cpp-mlp/PIM/mlp.cpp:1]: Missing file header and license info

File: misc-bench/cpp-mlp/baselines/mlp.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
+import sys
```
Comment (deyuan) [inline misc-bench/cpp-mlp/baselines/mlp.py:1]: Missing file header

**Review by deyuan — COMMENTED** [review]
Hi @kmd2zjw I've added a few comments. Could you take a look?

**Comment by kmd2zjw** [comment]
This is an outdated PR, but I implemented most of your suggestions. The finalized MLP needs to implement batched GEMM, which I'm still waiting to be fully finished and pushed. The part that isn't finished is listed here in a comment in the batched gemm pim file:

// Concatenate matrices. Needs to be done once
  std::vector<std::vector<int>> batchedSrcMatA;
  for (uint64_t i = 0; i < colA; ++i)
  {
    std::vector<int> temp;
    for (uint64_t j = 0; j < batchSize; ++j) {
      // TODO: Add replicate a pattern API in the simulator
      for (uint64_t k = 0; k < row; ++k) {
        temp.push_back(srcMatrixAT[i][k]);
      }
    }
    batchedSrcMatA.push_back(temp);
  }

**Pull request closed without merging**